include/lbann/layers/transform/dist_SOB.hpp

```diff
@@ -0,0 +1,223 @@
+    // }
+
+  
+
+  }
+
+  void fp_setup_outputs(El::Int mini_batch_size) override {
+
+    if (this->get_num_children() < 1) { return; }
+    // Determine distributed matrix alignment
+    const bool align_outputs = this->get_num_parents() > 0;
+    const auto& alignment_dist = (align_outputs ?
+                                  this->get_prev_activations().DistData() :
+                                  this->get_activations().DistData());
+
+
+    // Initialize output tensors
+    for (int i = 0; i < this->get_num_children(); ++i) {
+
+      auto& output = this->get_activations(i);
+      output.Empty(false);
+      if (align_outputs &&  this->get_parallel_strategy().enable_subgraph==0) { output.AlignWith(alignment_dist); }
+      output.Resize(this->get_output_size(i), mini_batch_size);
+      }
+
+  }
+
+  void bp_setup_gradient_wrt_inputs(El::Int mini_batch_size) override {
+    int tag= 0 ;
+    const auto& gradient_wrt_output = this->get_prev_error_signals();
+    
+
+    if(this->is_subgraph_parallelism_enabled() && this->get_parallel_strategy().enable_subgraph==1)
+    {
+      auto subgrid_tags = (*this->parent_tags);
+      
+      if(this->get_communication_flag())
+      //If vector copy is enable, broadcast the gradients from parent grid to multiple subgrids
+      {
+        auto const* ptr_gradient = dynamic_cast<El::DistMatrix<TensorDataType, El::STAR  , El::VC, El::ELEMENT, Dev> const*>(&gradient_wrt_output);
+        El::copy::TranslateBetweenGridsBroadcast<TensorDataType,Dev,Dev>(*ptr_gradient,this->get_branch_tag_input_vector(),this->get_subgrid_comm(),syncSubGridCommunication);
+
+      }
+      else if(this->get_communication_flag()==1)
+      {
+        auto const* ptr_gradient = dynamic_cast<El::DistMatrix<TensorDataType, El::STAR  , El::VC, El::ELEMENT, Dev> const*>(&gradient_wrt_output);
+        El::copy::TranslateBetweenGridsBroadcast<TensorDataType,Dev,Dev>(*ptr_gradient,this->get_branch_tag_input_vector());
+
+
+      }
+      else{
+        for(int i = 0; i < this->num_spliting_groups; i++)
+        {
+          
+          El::Copy( gradient_wrt_output, this->get_branch_tag_input(i));
+        }
+
+      } //end vector copy condition
+      
+      
+
+      for (int i = 0; i < this->get_num_parents(); ++i) {
+        tag = subgrid_tags[i];
+        
+        El::LockedView(this->get_error_signals(i), this->get_branch_tag_input(tag));
+        
+      }
+      
+    } 
+    else
+    {
+      for (int i = 0; i < this->get_num_parents(); ++i) {
+        
+        El::LockedView(this->get_error_signals(i), gradient_wrt_output);
+
+      }
+    }
+
+
+  }
+
+  void bp_compute() override {}
+
+
+};
+
+
+
+LBANN_DEFINE_LAYER_BUILDER(dist_SOB);
+
+#ifndef LBANN_DIST_SOB_LAYER_INSTANTIATE
+#define PROTO_DEVICE(T, Device) \
+  extern template class dist_SOB_layer<T, data_layout::DATA_PARALLEL, Device>; \
+  extern template class dist_SOB_layer<T, data_layout::MODEL_PARALLEL, Device>
```

Comment:
Model-parallel case is not supported.
```suggestion
  extern template class dist_SOB_layer<T, data_layout::DATA_PARALLEL, Device>;
```